source/common/upstream/cds_api_impl.cc

```diff
@@ -54,7 +71,14 @@ void CdsApiImpl::onConfigUpdate(const ResourceVector& resources, const std::stri
   for (auto& cluster : resources) {
     const std::string cluster_name = cluster.name();
     clusters_to_remove.erase(cluster_name);
-    if (cm_.addOrUpdateCluster(cluster, version_info)) {
+    if (cm_.addOrUpdateCluster(cluster, version_info, [this](auto, auto state) {
+          if (cm_.warmingClusterCount() == 0) {
```

Comment:
I'm still wondering about the case where `addOrUpdateCluster` on the first iteration comes back immediately Finished, we have 0 warming clusters and resume, and then on the next cluster it decides to warm, but we've already unpaused.

I think there is an implicit expectation that `addOrUpdateCluster()` always defers and never returns immediately if there is an immediate result of the cluster warm. This may be true today, but in the future we have discussed doing things like having EDS cache earlier results to avoid re-fetching hosts for a cluster undergoing updates. In this situation, it would probably warm synchronously.

WDYT? I don't want to over-complicate this, but it would be good to be robust to this future direction.

Reply:
Thanks for clarifying this. I think we should play safe in that case and make sure logic will work correctly in case we add EDS caching in the future. I changed the warming callback to check for both pause and resume conditions. I removed the auxiliary `pauseCdsWhileWarming()` and `resumeCdsAfterWarming()` as I think the code is now cleaner without them.

Also added a unit test to verify behavior when cluster is immediately warmed.